### PR TITLE
Allow settings migration from packet tunnel

### PR DESCRIPTION
**Pull request by buggmagnet**
This PR adds the ability for the PacketTunnel settings to perform settings migration.
The net benefit of this is that users that have automatic app updates turned on won't have to manually disconnect and reconnect from the VPN when the updates itself, the upgrade should be transparent for them.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)

### CODE REVIEW

**Review by pinkisemils — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @buggmagnet)

___
*[`ios/MullvadSettings/MigrationManager.swift` line 39 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8b-bGyBLWX2BsAJq9p:-O8b-bGyBLWX2BsAJq9q:b-ngnz3b) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadSettings/MigrationManager.swift#L39)):*
> ```Swift
>     /// In order to avoid migration happening from both the VPN and the host processes at the same time,
>     /// a non existant file path is used as a lock is used to synchronize access between the processes.
>     /// This file is accessed by `NSFileCoordinator` in order to prevent multiple processes accessing at the same time.
> ```

If we can lock a file path, why not use the path to the settings file instead? No need for locking unrelated files.

**Review by buggmagnet — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @pinkisemils)

___
*[`ios/MullvadSettings/MigrationManager.swift` line 39 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8b-bGyBLWX2BsAJq9p:-O8b06Di4oTspFQsoHXl:bysg6cf) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadSettings/MigrationManager.swift#L39)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

If we can lock a file path, why not use the path to the settings file instead? No need for locking unrelated files.
</blockquote></details>

Because the settings are not stored in a file, but in the keychain itself

**Review by pinkisemils — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @buggmagnet)

___
*[`ios/MullvadSettings/MigrationManager.swift` line 39 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8b-bGyBLWX2BsAJq9p:-O8b0n-Q0KRPIPbrLaO5:b-njaagj) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadSettings/MigrationManager.swift#L39)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Because the settings are not stored in a file, but in the keychain itself
</blockquote></details>

You are correct.  But in that case, shouldn't this lock be acquired anytime the settings are read?

**Review by pinkisemils — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @buggmagnet)

___
*[`ios/MullvadSettings/MigrationManager.swift` line 39 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8b-bGyBLWX2BsAJq9p:-O8bJJD0EoUQDKXpm5kH:bvcfzjs) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadSettings/MigrationManager.swift#L39)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

You are correct.  But in that case, shouldn't this lock be acquired anytime the settings are read?
</blockquote></details>

No, we do not need to do that if we ensure the migration is done when the UI process starts. This is sound now. Great work!
___
*[`ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift` line 193 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8bJU2G9t_x-4acPwQ-:-O8bJU2G9t_x-4acPwQ0:b-7k4sds) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift#L193)):*
> ```Swift
>                         )
>                     DispatchQueue.main.asyncAfter(deadline: .now() + delay.timeInterval) { [unowned self] in
>                         performSettingsMigration()
> ```

If this does not block the startup, what does the packet tunnel read from the settings? Would it not fail to deserialize or read garbage constraints? Does it make sense to have the packet tunnel not block it's startup on this?

**Review by buggmagnet — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @pinkisemils)

___
*[`ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift` line 193 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8bJU2G9t_x-4acPwQ-:-O8bMCTHB23f2amEAw-_:b-30c2fv) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift#L193)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

Packet tunnel will get killed in 60 seconds if doesn't start anyway.
Our best hope is that we enter the blocked state until the user starts using their phones, at which point we're at the mercy of the retry logic.

Do keep in mind that this scenario is *extremely unlikely*. It would mean the phone both had a reboot AND and an app update before the user unlocked their phone after the reboot. 

I would not worry about those scenarios.
</blockquote></details>

The other scenario is that the phone was such in a bad state that we were not able to upgrade the settings. 
Our upgrade process right now is basically infaillible (since we removed the migration that did networking calls after SettingsV1).

**Review by rablador — CHANGES_REQUESTED** [review]
> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: all files reviewed, 5 unresolved discussions (waiting on @buggmagnet and @pinkisemils)

___
*[`ios/MullvadREST/RetryStrategy/RetryStrategy.swift` line 83 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8eu0_87o_JAt5ilLRI:-O8eu0_87o_JAt5ilLRJ:b-j64ttu) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadREST/RetryStrategy/RetryStrategy.swift#L83)):*
> ```Swift
> 
>         public static var failedMigrationRecovery = RetryStrategy(
>             maxRetryCount: .max,
> ```

Are we going to retry this basically forever?
___
*[`ios/MullvadSettings/MigrationManager.swift` line 38 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8evG1E-fa193HbFXPD:-O8evG1E-fa193HbFXPE:b-muivs3) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadSettings/MigrationManager.swift#L38)):*
> ```Swift
>     ///
>     /// In order to avoid migration happening from both the VPN and the host processes at the same time,
>     /// a non existant file path is used as a lock is used to synchronize access between the processes.
> ```

Nit: "is used as a lock is used to"
___
*[`ios/MullvadVPNTests/MullvadTypes/FileCacheTests.swift` line 10 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8ex0bk4FcmEp8w1xBe:-O8ex0bk4FcmEp8w1xBf:bwgzvz9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadVPNTests/MullvadTypes/FileCacheTests.swift#L10)):*
> ```Swift
> 
> import Foundation
> @testable import MullvadMockData
> ```

Added by mistake?
___
*[`ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift` line 193 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8bJU2G9t_x-4acPwQ-:-O8exj1aB-Ji9EnCeT-m:b-oc61z3) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift#L193)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

The other scenario is that the phone was such in a bad state that we were not able to upgrade the settings. 
Our upgrade process right now is basically infaillible (since we removed the migration that did networking calls after SettingsV1).
</blockquote></details>

So basically, before migration happens, the tunnel is started and (possibly) enters blocked state. Then migration finishes and hopefully succeeds. The tunnel reconnection timer fires at some point and user leaves blocked state?

**Review by buggmagnet — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: all files reviewed, 5 unresolved discussions (waiting on @pinkisemils and @rablador)

___
*[`ios/MullvadREST/RetryStrategy/RetryStrategy.swift` line 83 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8eu0_87o_JAt5ilLRI:-O8f69qi8wZn9KtLGFPV:b-bxq1ft) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadREST/RetryStrategy/RetryStrategy.swift#L83)):*
That's what the ticket asks for:
> The packet tunnel must be able to handle a reboot - when the keychain is locked. It should retry with a backoff.
___
*[`ios/MullvadSettings/MigrationManager.swift` line 38 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8evG1E-fa193HbFXPD:-O8f6ckC9o9EYpDwbfcs:b-1akjx8) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadSettings/MigrationManager.swift#L38)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

Nit: "is used as a lock is used to"
</blockquote></details>

I reread this 3 times and yet... nice catch ! 
___
*[`ios/MullvadVPNTests/MullvadTypes/FileCacheTests.swift` line 10 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8ex0bk4FcmEp8w1xBe:-O8f6lmV6oXRcV8AnJgd:b-amoa76) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadVPNTests/MullvadTypes/FileCacheTests.swift#L10)):*
<details><summary><i>Previously, rablador (Jon Petersson) wrote…</i></summary><blockquote>

Added by mistake?
</blockquote></details>

yes, leftover of another approach I had tried where I wanted to add a test here, will delete. 
___
*[`ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift` line 193 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8bJU2G9t_x-4acPwQ-:-O8f6s3_7w8NymihKsZh:b-gzdkwa) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift#L193)):*
I want to emphasize that the odds of this failing are astronomically low for 2 reasons:
- The `upgradeToNextVersion` function cannot fail. i.e. the only failure mode is keychain settings have been badly corrupted (or we messed up our next `TunnelSettings` by adding fields that do not deserialize), at which point we cannot guarantee the app will work reliably at all. If this fails, upgrading from the UI process will end in the same result, and the user will have to re-install the app in any case
- The scenario in which we cannot read from the settings because the keychain is still locked only happens if the user's device rebooted without the user interacting with the device (since they would otherwise unlock the device to use it). We could actually remove this restriction by using `kSecAttrAccessibleAlways` instead. If we're really worried about other apps reading our keychain entries (or reading them from a backup) there are better solutions around. 

> before migration happens, the tunnel is started and (possibly) enters blocked state.

Migration is not asynchronous, the tunnel won't start before migration is done, therefore we cannot enter the blocked state _because_ we are waiting on migration. It either succeeds and we start the tunnel, or it fails

**Review by rablador — COMMENTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: all files reviewed, 3 unresolved discussions (waiting on @buggmagnet and @pinkisemils)

___
*[`ios/MullvadREST/RetryStrategy/RetryStrategy.swift` line 83 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8eu0_87o_JAt5ilLRI:-O8fBm1i1mBLzA9aiUr5:bk3b2kh) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/MullvadREST/RetryStrategy/RetryStrategy.swift#L83)):*
<details><summary><i>Previously, buggmagnet wrote…</i></summary><blockquote>

That's what the ticket asks for:
> The packet tunnel must be able to handle a reboot - when the keychain is locked. It should retry with a backoff.
</blockquote></details>

Still, forever...?
___
*[`ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift` line 193 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-O8bJU2G9t_x-4acPwQ-:-O8fBLKk13qgNCuT36uA:b-ewbwbv) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/ff36ba9521c08aa894c974ccab2d0fb26c5f2c27/ios/PacketTunnel/PacketTunnelProvider/PacketTunnelProvider.swift#L193)):*
> Migration is not asynchronous, the tunnel won't start before migration is done

Aha, then all is good I think. Checked the Linear issue just now and one of the acceptace criteria also states that blocked state should not happen.

**Review by rablador — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-:-O8fuD7xESNWEu9eXuMN:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @buggmagnet and @pinkisemils)

**Review by rablador — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938)* status: all files reviewed, 2 unresolved discussions (waiting on @pinkisemils)

**Review by pinkisemils — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/6938#-:-O8kNJ4_4Bf4sjh-3xQz:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved
